plugin/evm/gossip_mempool.go

```diff
@@ -0,0 +1,217 @@
+	lock    sync.RWMutex
+}
+
+func (g *GossipAtomicMempool) Add(tx *GossipAtomicTx) (bool, error) {
+	ok, err := g.mempool.AddTx(tx.Tx)
+	if err != nil {
+		if !tx.Local {
+			// unlike local txs, invalid remote txs are recorded as discarded
+			// so that they won't be requested again
+			txID := tx.Tx.ID()
+			g.mempool.discardedTxs.Put(txID, tx.Tx)
+			log.Debug("failed to issue remote tx to mempool",
+				"txID", txID,
+				"err", err,
+			)
+		}
+		return false, err
+	}
+
+	if !ok {
+		return false, nil
+	}
+
+	g.lock.Lock()
+	defer g.lock.Unlock()
+
+	g.bloom.Add(gossip.NewHasher(tx.GetID()))
+	g.bloom, _ = gossip.ResetBloomFilterIfNeeded(g.bloom, gossip.DefaultBloomMaxFilledRatio)
+
+	return true, nil
+}
+
+func (g *GossipAtomicMempool) Get(filter func(tx *GossipAtomicTx) bool) []*GossipAtomicTx {
+	f := func(tx *Tx) bool {
+		return filter(&GossipAtomicTx{
+			Tx: tx,
+		})
+	}
+	txs := g.mempool.GetTxs(f)
+	gossipTxs := make([]*GossipAtomicTx, 0, len(txs))
+	for _, tx := range txs {
+		gossipTxs = append(gossipTxs, &GossipAtomicTx{
+			Tx: tx,
+		})
+	}
+
+	return gossipTxs
+}
+
+func (g *GossipAtomicMempool) GetBloomFilter() ([]byte, error) {
+	g.lock.RLock()
+	defer g.lock.RUnlock()
+
+	return g.bloom.MarshalBinary()
+}
+
+type GossipAtomicTx struct {
+	Tx    *Tx
+	Local bool
+}
+
+func (tx *GossipAtomicTx) GetID() ids.ID {
+	return tx.Tx.ID()
+}
+
+func (tx *GossipAtomicTx) Marshal() ([]byte, error) {
+	return Codec.Marshal(message.Version, tx.Tx)
+}
+
+func (tx *GossipAtomicTx) Unmarshal(bytes []byte) error {
+	tx.Tx = &Tx{}
+	_, err := Codec.Unmarshal(bytes, tx.Tx)
+
+	return err
+}
+
+func NewGossipEthTxPool(mempool *txpool.TxPool) (*GossipEthTxPool, error) {
+	bloom, err := bloomfilter.New(gossip.DefaultBloomM, gossip.DefaultBloomK)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize bloom filter: %w", err)
+	}
+
+	g := &GossipEthTxPool{
+		mempool:    mempool,
+		pendingTxs: make(chan core.NewTxsEvent),
+		bloom:      bloom,
+	}
+	return g, nil
+}
+
+type GossipEthTxPool struct {
+	mempool    *txpool.TxPool
+	pendingTxs chan core.NewTxsEvent
+
+	bloom *bloomfilter.Filter
+	lock  sync.RWMutex
+}
+
+func (g *GossipEthTxPool) Subscribe(shutdownChan chan struct{}, shutdownWg *sync.WaitGroup) {
+	defer shutdownWg.Done()
+	g.mempool.SubscribeNewTxsEvent(g.pendingTxs)
+
+	for {
+		select {
+		case <-shutdownChan:
+			log.Debug("shutting down subscription")
+			return
+		case tx := <-g.pendingTxs:
+			g.lock.Lock()
+			for _, tx := range tx.Txs {
+				g.bloom.Add(gossip.NewHasher(ids.ID(tx.Hash())))
+				g.bloom, _ = gossip.ResetBloomFilterIfNeeded(g.bloom, gossip.DefaultBloomMaxFilledRatio)
+			}
+			g.lock.Unlock()
+		}
+	}
+}
+
+// Add enqueues the transaction to the mempool. Subscribe should be called
+// to receive an event if tx is actually added to the mempool or not.
+func (g *GossipEthTxPool) Add(tx *GossipEthTx) (bool, error) {
+	err := g.mempool.AddRemotes([]*types.Transaction{tx.Tx})[0]
```

Comment:
I am sure this is nil guarded at a higher level but this does raise my eyebrows :)

Reply:
Yeah this function always returns a slice of errors of len(txs) so we're guaranteed to get a slice of errors of length 1 here